cmd/kubeadm/app/util/etcd/etcd_test.go

```diff
@@ -494,3 +495,221 @@ func TestClient_GetMemberID(t *testing.T) {
+	type fields struct {
+		Endpoints     []string
+		newEtcdClient func(endpoints []string) (etcdClient, error)
+	}
+	tests := []struct {
+		name     string
+		fields   fields
+		memberID uint64
+		want     bool
+	}{
+		{
+			name: "The specified member is not leader",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{
+						members: []*pb.Member{
+							{
+								ID:   1,
+								Name: "member1",
+								PeerURLs: []string{
+									"https://member1:2380",
+								},
+							},
+							{
+								ID:   2,
+								Name: "member2",
+								PeerURLs: []string{
+									"https://member2:2380",
+								},
+							},
+						},
+					}
+					return f, nil
+				},
+			},
+			memberID: 1,
+			want:     false,
+		},
+		{
+			name: "The specified member is leader",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{
+						members: []*pb.Member{
+							{
+								ID:   1,
+								Name: "member1",
+								PeerURLs: []string{
+									"https://member1:2380",
+								},
+								IsLearner: true,
+							},
+							{
+								ID:   2,
+								Name: "member2",
+								PeerURLs: []string{
+									"https://member2:2380",
+								},
+							},
+						},
+					}
+					return f, nil
+				},
+			},
+			memberID: 1,
+			want:     true,
+		},
+		{
+			name: "The specified member is not exist",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{
+						members: []*pb.Member{
+							{
+								ID:   1,
+								Name: "member1",
+								PeerURLs: []string{
+									"https://member1:2380",
+								},
+								IsLearner: true,
+							},
+							{
+								ID:   2,
+								Name: "member2",
+								PeerURLs: []string{
+									"https://member2:2380",
+								},
+							},
+						},
+					}
+					return f, nil
+				},
+			},
+			memberID: 3,
+			want:     false,
+		},
+		{
+			name: "Leader member ID is empty",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{
+						members: []*pb.Member{
+							{
+								Name: "member2",
+								PeerURLs: []string{
+									"https://member2:2380",
+								},
+								IsLearner: true,
+							},
+							{
+								ID:   2,
+								Name: "member2",
+								PeerURLs: []string{
+									"https://member2:2380",
+								},
+							},
+						},
+					}
+					return f, nil
+				},
+			},
+			want: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := &Client{
+				Endpoints:     tt.fields.Endpoints,
+				newEtcdClient: tt.fields.newEtcdClient,
+			}
+			got, err := c.isLearner(tt.memberID)
+			if got != tt.want {
+				t.Errorf("isLearner() = %v, want %v", got, tt.want)
+			}
+			assert.NoError(t, err)
```

Comment:
can we simulate error cases where `err != nil` ?

Reply:
to enable error testing, here we can pass a custom listMembersFunc
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/etcd/etcd.go#L482-L483

```
func (c *Client) isLearner(memberID uint64, listMemberFunc (*clientv3.MemberListResponse, error)) { ) (bool, error) {
    if listMemberFunc == nil {
        istMemberFunc = c.listMembers()
    }
    resp, err := listMembersFunc()
```

then during unit tests you can feed a custom listMembersFunc that can return error or just return a list of members.
we are already testing the actual listMembers() so no need to test it again.

Reply:
alternatively, functions can be added to the Client struct.
the same way `newEtcdClient: tt.fields.newEtcdClient,` is done.